src/ninja.cc

```diff
@@ -505,6 +507,220 @@ int NinjaMain::ToolDeps(const Options* options, int argc, char** argv) {
+  set<Node*> generated_nodes_;
+  set<const Rule*> generator_rules_;
+  int missing_dep_path_count_;
+  set<Node*> nodes_missing_gen_;
+  string abs_gen_dir_;
+};
+
+MissingDependencyScanner::MissingDependencyScanner(
+    DepsLog* deps_log, State* state, DiskInterface* disk_interface,
+    const string& gen_dir)
+    : deps_log_(deps_log), state_(state), disk_interface_(disk_interface),
+      missing_dep_path_count_(0) {
+  if (!gen_dir.empty()) {
+    char buf[PATH_MAX];
+    if (getcwd(buf, PATH_MAX) == NULL)
+      Fatal("getcwd failed");
+    abs_gen_dir_ = buf;
+    abs_gen_dir_.append("/");
+    abs_gen_dir_.append(gen_dir);
+    if (!realpath(abs_gen_dir_.c_str(), buf))
+      Fatal("realpath for %s failed", abs_gen_dir_.c_str());
+    abs_gen_dir_ = buf;
+  }
+}
+
+bool MissingDependencyScanner::IsNodeInGenDir(Node* node) {
+  if (abs_gen_dir_.empty())
+    return false;
+  string path = node->path();
+  char abs_path_buf[PATH_MAX];
+  char* rp = realpath(path.c_str(), abs_path_buf);
+  return (rp && (strlen(rp) > abs_gen_dir_.size())) &&
+         (memcmp(rp, abs_gen_dir_.c_str(), abs_gen_dir_.size()) == 0);
+}
+
+void MissingDependencyScanner::ProcessNode(Node* node) {
+  if (!node)
+    return;
+  Edge* edge = node->in_edge();
+  if (!edge)
+    return;
+  if (!seen_.insert(node).second)
+    return;
+
+  for (vector<Node*>::iterator in = edge->inputs_.begin();
+       in != edge->inputs_.end(); ++in) {
+    ProcessNode(*in);
+  }
+
+  string deps_type = edge->GetBinding("deps");
+  if (!deps_type.empty()) {
+    DepsLog::Deps* deps = deps_log_->GetDeps(node);
+    if (deps)
+      ProcessNodeDeps(node, deps->nodes, deps->node_count);
+  } else {
+    ImplicitDepLoader dep_loader(state_, deps_log_, disk_interface_);
+    vector<Node*> depfile_deps;
+    string depfile = edge->GetUnescapedDepfile();
+    if (depfile.empty())
+      return;
+    string err;
+    dep_loader.LoadDepfileDeps(edge, depfile, &depfile_deps, &err);
+    if (!depfile_deps.empty())
+      ProcessNodeDeps(node, &depfile_deps[0], depfile_deps.size());
+  }
+}
+
+void MissingDependencyScanner::ProcessNodeDeps(Node* node, Node** dep_nodes,
+                                               int dep_nodes_count) {
+  Edge* edge = node->in_edge();
+  set<Edge*> deplog_edges;
+  vector<Node*> deplog_nodes_without_generator;
+  for (int i = 0; i < dep_nodes_count; ++i) {
+    Node* deplog_node = dep_nodes[i];
+    Edge* deplog_edge = deplog_node->in_edge();
+    if (deplog_edge) {
+      deplog_edges.insert(deplog_edge);
+    } else if (IsNodeInGenDir(deplog_node)) {
+      nodes_missing_gen_.insert(deplog_node);
+      deplog_nodes_without_generator.push_back(deplog_node);
+    }
+  }
+  vector<Edge*> missing_deps;
+  for (set<Edge*>::iterator de = deplog_edges.begin(); de != deplog_edges.end();
+       ++de) {
+    if (!state_->PathExistsBetween(*de, edge)) {
+      missing_deps.push_back(*de);
+    }
+  }
+
+  if (!deplog_nodes_without_generator.empty()) {
```

Comment:
Feel redundant.

Reply:
I think this is also obsolete after code move to src/missing_deps.cc and -g tool removal